src/middleware/origin.rs

```diff
@@ -0,0 +1,81 @@
+use hyper::status::StatusCode;
+use iron::middleware::BeforeMiddleware;
+use iron::modifiers;
+use iron::{IronError, IronResult, Request};
+use std::error::Error;
+use std::fmt;
+use url::Origin::{Opaque, Tuple};
+use url;
+
+/// Middleware that redirects requests to the broker's canonical public_url.
+pub struct EnforceOrigin {
+    origin: url::Origin,
+}
+
+impl EnforceOrigin {
+    pub fn new(uri: &str) -> EnforceOrigin {
+        let origin =
+            url::Url::parse(uri).expect(format!("unable to parse uri: {}", uri).as_str()).origin();
+
+        EnforceOrigin { origin: origin }
+    }
+}
+
+impl BeforeMiddleware for EnforceOrigin {
+    fn before(&self, req: &mut Request) -> IronResult<()> {
+        let mut uri = req.url.clone().into_generic_url();
+
+        // If the request and canonical origins match, everything is fine.
+        if self.origin == uri.origin() {
+            return Ok(());
+        }
+
+        // Otherwise, extract the configured, canonical origin...
+        let (scheme, host, port) = match self.origin {
+            Tuple(ref scheme, ref host, ref port) => (scheme, host, port),
+            Opaque(_) => {
+                return Err(IronError::new(OriginError::Opaque, StatusCode::InternalServerError))
+            }
+        };
+
+        // ...mutate the request URI to point to the canonical origin...
+        if uri.set_scheme(scheme).is_err() ||
+           uri.set_host(Some(&format!("{}", host))).is_err() ||
+           uri.set_port(Some(*port)).is_err() {
+            return Err(IronError::new(OriginError::Mutation, StatusCode::InternalServerError));
+        }
+
+        // ...and redirect to it.
+        let status = StatusCode::TemporaryRedirect;
+        let location = modifiers::RedirectRaw(uri.to_string());
+
+        Err(IronError::new(OriginError::Mismatch, (status, location)))
```

Comment:
Downside here is that an `Err` gets logged, right? I've avoided logging anything for input errors. I'd prefer an AroundMiddleware here, if possible.